examples/softirq-latency.bpf.c

```diff
@@ -0,0 +1,136 @@
+#include <vmlinux.h>
+#include <bpf/bpf_tracing.h>
+#include "bits.bpf.h"
+#include "maps.bpf.h"
+
+// Loosely based on https://github.com/xdp-project/xdp-project/blob/master/areas/latency/softirq_net_latency.bt
+
+// 20 buckets for latency, max range is 0.5s .. 1.0s
+#define MAX_LATENCY_SLOT 21
+
+struct softirq_latency_key_t {
+    u32 kind;
+    u64 bucket;
+};
+
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+    __uint(max_entries, NR_SOFTIRQS);
+    __type(key, u32);
+    __type(value, u64);
+} softirq_raised_total SEC(".maps");
+
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+    __uint(max_entries, NR_SOFTIRQS);
+    __type(key, u32);
+    __type(value, u64);
+} softirq_serviced_total SEC(".maps");
+
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+    __uint(max_entries, NR_SOFTIRQS);
+    __type(key, u32);
+    __type(value, u64);
+} softirq_raise_timestamp SEC(".maps");
+
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+    __uint(max_entries, NR_SOFTIRQS);
+    __type(key, u32);
+    __type(value, u64);
+} softirq_entry_timestamp SEC(".maps");
+
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_HASH);
+    __uint(max_entries, (MAX_LATENCY_SLOT + 1) * NR_SOFTIRQS);
+    __type(key, struct softirq_latency_key_t);
+    __type(value, u64);
+} softirq_entry_latency_seconds SEC(".maps");
+
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_HASH);
+    __uint(max_entries, (MAX_LATENCY_SLOT + 1) * NR_SOFTIRQS);
+    __type(key, struct softirq_latency_key_t);
+    __type(value, u64);
+} softirq_service_latency_seconds SEC(".maps");
+
+SEC("tp_btf/softirq_raise")
+int BPF_PROG(softirq_raise, unsigned int vec_nr)
+{
+    u64 *existing_ts_ptr, *raised_total_ptr, ts = bpf_ktime_get_ns();
+
+    read_array_ptr(&softirq_raised_total, &vec_nr, raised_total_ptr);
+    *raised_total_ptr += 1;
+
+    read_array_ptr(&softirq_raise_timestamp, &vec_nr, existing_ts_ptr);
+
+    // Set the timestamp only if it is not set, so that we always measure the oldest non-entered raise
+    if (!*existing_ts_ptr) {
+        *existing_ts_ptr = ts;
+    }
+
+    return 0;
+}
+
+SEC("tp_btf/softirq_entry")
+int BPF_PROG(softirq_entry, unsigned int vec_nr)
+{
+    u64 delta_us, *raise_ts_ptr, *existing_entry_ts_ptr, *serviced_total_ptr, ts = bpf_ktime_get_ns();
+    struct softirq_latency_key_t key = {};
+
+    read_array_ptr(&softirq_serviced_total, &vec_nr, serviced_total_ptr);
+    *serviced_total_ptr += 1;
+
+    read_array_ptr(&softirq_raise_timestamp, &vec_nr, raise_ts_ptr);
+
+    // Interrupt was re-rased after ts was obtained, resulting in negative duration
+    if (*raise_ts_ptr > ts) {
```

Comment:
This happens mostly for `TASKLET_SOFTIRQ` and sometimes for `TASKLET_SOFTIRQ`:

```
<...>-27627   [043] d.s3. 46693.775555: bpf_trace_printk: re-raise after 1500ns [6]
          <idle>-0       [035] d.s4. 46693.884661: bpf_trace_printk: re-raise after 2040ns [6]
          <idle>-0       [061] d.s4. 46694.595380: bpf_trace_printk: re-raise after 1650ns [6]
          <idle>-0       [047] d.s4. 46694.922268: bpf_trace_printk: re-raise after 1650ns [6]
          <idle>-0       [001] d.s4. 46694.942947: bpf_trace_printk: re-raise after 1120ns [6]
           <...>-1445425 [041] d.s4. 46694.951062: bpf_trace_printk: re-raise after 1450ns [6]
          <idle>-0       [071] d.s4. 46695.262426: bpf_trace_printk: re-raise after 840ns [7]
 conduit-watcher-2620846 [039] d.s3. 46695.998234: bpf_trace_printk: re-raise after 1430ns [6]
          <idle>-0       [003] d.s4. 46696.024913: bpf_trace_printk: re-raise after 2090ns [6]
          <idle>-0       [029] d.s4. 46696.244701: bpf_trace_printk: re-raise after 1150ns [6]
```

The working theory is that there can be a raise after entry started but before this condition is checked.

There aren't too many of these, so we mostly ignore them.